pkg/controller/direct/securesourcemanager/securesourcemanagerinstance_controller.go

```diff
@@ -0,0 +1,280 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package securesourcemanager
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	api "cloud.google.com/go/securesourcemanager/apiv1"
+	pb "cloud.google.com/go/securesourcemanager/apiv1/securesourcemanagerpb"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/apis/refs/refresolver"
+	refs "github.com/GoogleCloudPlatform/k8s-config-connector/apis/refs/v1beta1"
+	krm "github.com/GoogleCloudPlatform/k8s-config-connector/apis/securesourcemanager/v1alpha1"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/config"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct/directbase"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct/registry"
+	"google.golang.org/protobuf/proto"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func init() {
+	registry.RegisterModel(krm.SecureSourceManagerInstanceGVK, newSecureSourceManagerInstanceModel)
+}
+
+func newSecureSourceManagerInstanceModel(ctx context.Context, config *config.ControllerConfig) (directbase.Model, error) {
+	return &SecureSourceManagerInstanceModel{config: config}, nil
+}
+
+type SecureSourceManagerInstanceModel struct {
+	config *config.ControllerConfig
+}
+
+// model implements the Model interface.
+var _ directbase.Model = &SecureSourceManagerInstanceModel{}
+
+type instanceAdapter struct {
+	projectID  string
+	location   string
+	resourceID string
```

Comment:
Mind adding a todo about using externalRef instead of the resourceID? Just in case we forget it.

Reply:
oh, because https://github.com/GoogleCloudPlatform/k8s-config-connector/pull/2370/files is merged after you run the controller template. (was wondering why this controller contains many out of date changes).